src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,484 @@
+
+/-- Given the set `i`, if `i` is not negative, `find_exists_one_div_lt s i` is the
+least natural number `n` such that `exists_one_div_lt s i n`, otherwise, it returns 0. -/
+private def find_exists_one_div_lt (s : signed_measure α) (i : set α) : ℕ :=
+if hi : ¬ s ≤[i] 0 then nat.find (exists_nat_one_div_lt_measure_of_not_negative hi) else 0
+
+private lemma find_exists_one_div_lt_spec (hi : ¬ s ≤[i] 0) :
+  exists_one_div_lt s i (find_exists_one_div_lt s i) :=
+begin
+  rw [find_exists_one_div_lt, dif_pos hi],
+  convert nat.find_spec _,
+end
+
+private lemma find_exists_one_div_lt_min (hi : ¬ s ≤[i] 0) {m : ℕ}
+  (hm : m < find_exists_one_div_lt s i) : ¬ exists_one_div_lt s i m :=
+begin
+  rw [find_exists_one_div_lt, dif_pos hi] at hm,
+  exact nat.find_min _ hm
+end
+
+/-- Given the set `i`, if `i` is not negative, `some_exists_one_div_lt` chooses the set
+`k` from `exists_one_div_lt s i (find_exists_one_div_lt s i)`, otherwise, it returns the
+empty set. -/
+private def some_exists_one_div_lt (s : signed_measure α) (i : set α) : set α :=
+if hi : ¬ s ≤[i] 0 then classical.some (find_exists_one_div_lt_spec hi) else ∅
+
+private lemma some_exists_one_div_lt_spec (hi : ¬ s ≤[i] 0) :
+  (some_exists_one_div_lt s i) ⊆ i ∧ measurable_set (some_exists_one_div_lt s i) ∧
+  (1 / (find_exists_one_div_lt s i + 1) : ℝ) < s (some_exists_one_div_lt s i) :=
+begin
+  rw [some_exists_one_div_lt, dif_pos hi],
+  exact classical.some_spec (find_exists_one_div_lt_spec hi),
+end
+
+private lemma some_exists_one_div_lt_subset : some_exists_one_div_lt s i ⊆ i :=
+begin
+  by_cases hi : ¬ s ≤[i] 0,
+  { exact let ⟨h, _⟩ := some_exists_one_div_lt_spec hi in h },
+  { rw [some_exists_one_div_lt, dif_neg hi],
+    exact set.empty_subset _ },
+end
+
+private lemma some_exists_one_div_lt_subset' : some_exists_one_div_lt s (i \ j) ⊆ i :=
+set.subset.trans some_exists_one_div_lt_subset (set.diff_subset _ _)
+
+private lemma some_exists_one_div_lt_measurable_set :
+  measurable_set (some_exists_one_div_lt s i) :=
+begin
+  by_cases hi : ¬ s ≤[i] 0,
+  { exact let ⟨_, h, _⟩ := some_exists_one_div_lt_spec hi in h },
+  { rw [some_exists_one_div_lt, dif_neg hi],
+    exact measurable_set.empty }
+end
+
+private lemma some_exists_one_div_lt_lt (hi : ¬ s ≤[i] 0) :
+  (1 / (find_exists_one_div_lt s i + 1) : ℝ) < s (some_exists_one_div_lt s i) :=
+let ⟨_, _, h⟩ := some_exists_one_div_lt_spec hi in h
+
+/-- Given the set `i`, `restrict_nonpos_seq s i` is the sequence of sets defined inductively where
+`restrict_nonpos_seq s i 0 = some_exists_one_div_lt s (i \ ∅)` and
+`restrict_nonpos_seq s i (n + 1) = some_exists_one_div_lt s (i \ ⋃ k ≤ n, restrict_nonpos_seq k)`.
+
+For each `n : ℕ`,`s (restrict_nonpos_seq s i n)` is close to maximal among all subsets of
+`i \ ⋃ k ≤ n, restrict_nonpos_seq s i k`. -/
+private def restrict_nonpos_seq (s : signed_measure α) (i : set α) : ℕ → set α
+| 0 := some_exists_one_div_lt s (i \ ∅) -- I used `i \ ∅` instead of `i` to simplify some proofs
```

Comment:
I'm surprised that this simplifies things!

Reply:
It only makes a difference since in some proofs it allows us to use the same proof for the base case and the succ case :)